pkg/auth/auth.go

```diff
@@ -182,30 +190,68 @@ func (a *Authorizer) setKeys(keys []interface{}) {
 	a.keyGuard.Unlock()
 }
 
-// Authorize extracts and verifies bearer tokens from a http.Request.
+// TokenMiddleware decodes the authentication token and passes the claims to the authorizer and to the context for logging.
+func (a *Authorizer) TokenMiddleware(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		claims, err := a.extractClaims(r)
+		if err == nil {
+			if !a.acceptedAudiences[claims.Audience] {
+				err = stacktrace.NewErrorWithCode(dsserr.Unauthenticated, "Invalid access token audience: %v", claims.Audience)
+			}
+		}
+
+		a.decodedClaims = &middlewareResult{
+			claims: claims,
+			err:    err,
+		}
+
+		var errMsg string
+		if err != nil {
+			//remove the stacktrace using the formatting specifier "%#s"
+			errMsg = fmt.Sprintf("%#s", err)
+		}
+
+		ctx := context.WithValue(r.Context(), logging.CtxKey("sub"), logging.CtxAuthValue{
+			Subject: claims.Subject,
+			ErrMsg:  errMsg,
+		})
+
+		handler.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// Authorize extracts and verifies bearer tokens from a http.Request after it was validated by the TokenMiddleware.
 func (a *Authorizer) Authorize(_ http.ResponseWriter, r *http.Request, authOptions []api.AuthorizationOption) api.AuthorizationResult {
-	keyClaims, err := a.ExtractClaims(r)
-	if err != nil {
-		return api.AuthorizationResult{Error: stacktrace.PropagateWithCode(err, dsserr.Unauthenticated, "Failed to extract claims from access token")}
+	if a.decodedClaims == nil {
```

Comment:
This should come from the context, and not form the `Authorizer`, c.f. other comment.